ui/guest/src/react/MoveModeZoneMenu.tsx

```diff
@@ -15,29 +15,89 @@
 import ArrowDownwardRoundedIcon from '@mui/icons-material/ArrowDownwardRounded';
 import ArrowUpwardRoundedIcon from '@mui/icons-material/ArrowUpwardRounded';
 import DeleteOutlineRoundedIcon from '@mui/icons-material/DeleteOutlineRounded';
 import UltraStyledIconButton from './UltraStyledIconButton';
 import HighlightOffRoundedIcon from '@mui/icons-material/HighlightOffRounded';
 import { Tooltip } from '@mui/material';
-import { useDispatch } from './GuestContext';
-import * as elementRegistry from '../classes/ElementRegistry';
-import DragGhostElement from './DragGhostElement';
+import * as contentController from '../classes/ContentController';
+import { clearAndListen$ } from '../store/subjects';
+import { startListening } from '../store/actions';
+import { compileDropZone, fromElement, getParentElementFromICEProps } from '../classes/ElementRegistry';
+import { ElementRecord } from '../models/InContextEditing';
 
 export interface MoveModeZoneMenuProps {
   [key: string]: any;
+  record: ElementRecord;
 }
 
 export function MoveModeZoneMenu(props: MoveModeZoneMenuProps) {
   const { record, dispatch } = props;
-  // const dispatch = useDispatch();
+  const {
+    modelId,
+    fieldId: [fieldId],
+    index
+  } = record;
+
+  const elementIndex = useMemo(() => {
+    if (typeof index === 'string') {
+      return parseInt(index.substr(index.lastIndexOf('.') + 1), 10);
+    }
+    return index;
+  }, [index]);
+
+  const dropzoneRecord = useMemo(
+    () => fromElement(getParentElementFromICEProps(modelId, fieldId, index)[0] as Element),
+    [fieldId, index, modelId]
+  );
+
+  const dropzoneChildrenLength = useMemo(
+    () => compileDropZone(dropzoneRecord.iceIds[0]).children.length,
+    [dropzoneRecord.iceIds[0]]
+  );
+
+  const isFirstItem = elementIndex === 0;
+  const isLastItem = elementIndex === dropzoneChildrenLength - 1;
+  const isOnlyItem = isFirstItem && isLastItem;
+
   // region callbacks
-  const onMoveUp = () => void 0;
-  const onMoveDown = () => void 0;
-  const onTrash = () => void 0;
-  const onCancel = () => void 0;
+  const clearAndStartListening = () => {
+    clearAndListen$.next();
+    dispatch(startListening());
+  };
+
+  const onMoveUp = () => {
+    contentController.sortItem(
+      dropzoneRecord.modelId,
+      dropzoneRecord.fieldId[0],
+      dropzoneRecord.fieldId[0].includes('.') ? `${dropzoneRecord.index}.${elementIndex}` : elementIndex,
+      dropzoneRecord.fieldId[0].includes('.') ? `${dropzoneRecord.index}.${elementIndex - 1}` : elementIndex - 1
+    );
+    clearAndStartListening();
+  };
+
+  const onMoveDown = () => {
+    contentController.sortItem(
+      dropzoneRecord.modelId,
+      dropzoneRecord.fieldId[0],
+      dropzoneRecord.fieldId[0].includes('.') ? `${dropzoneRecord.index}.${elementIndex}` : elementIndex,
+      dropzoneRecord.fieldId[0].includes('.') ? `${dropzoneRecord.index}.${elementIndex + 1}` : elementIndex + 1
+    );
+    clearAndStartListening();
+  };
```

Comment:
I think we should abstract this on the content controller: create `moveItemUp` & `moveItemDown` methods there and just invoke here. Method should take into account only items, index 0, last index and ignore updates in those cases.